### ISSUE TITLE
Device fails to import (Trac #106)

### ISSUE THREAD

**Issue by cfobel** [wheeler-microfluidics/microdrop#110]
When attempting to import the attached device, the following exception occurs:

```
(<type 'exceptions.AttributeError'>, AttributeError("'NoneType' object has no attribute 'right_point'",), <traceback object at 0xbe4216c>) {}
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/gui/dmf_device_controller.py", line 417, in on_import_dmf_device
    app.dmf_device = DmfDevice.load_svg(filename)
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/dmf_device.py", line 93, in load_svg
    dmf_device.init_body_group()
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/dmf_device.py", line 62, in init_body_group
    self.body_group = BodyGroup(self.path_group.paths)
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/body_group.py", line 33, in __init__
    self.add_path(name, geo_path)
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/body_group.py", line 42, in add_path
    triangulator = Triangulator(loop.verts)
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/seidel.py", line 213, in __init__
    self.process()
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/seidel.py", line 230, in process
    traps = self.query_graph.follow_edge(edge)
  File "/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/seidel.py", line 508, in follow_edge
    while(edge.q.x > trapezoids[-1].right_point.x):
```

Migrated from http://microfluidics.utoronto.ca/microdrop/ticket/106
```json
{
    "status": "closed", 
    "changetime": "2014-04-17T19:39:01", 
    "description": "When attempting to import the attached device, the following exception occurs:\n\n{{{\n(<type 'exceptions.AttributeError'>, AttributeError(\"'NoneType' object has no attribute 'right_point'\",), <traceback object at 0xbe4216c>) {}\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/gui/dmf_device_controller.py\", line 417, in on_import_dmf_device\n    app.dmf_device = DmfDevice.load_svg(filename)\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/dmf_device.py\", line 93, in load_svg\n    dmf_device.init_body_group()\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/dmf_device.py\", line 62, in init_body_group\n    self.body_group = BodyGroup(self.path_group.paths)\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/body_group.py\", line 33, in __init__\n    self.add_path(name, geo_path)\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/body_group.py\", line 42, in add_path\n    triangulator = Triangulator(loop.verts)\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/seidel.py\", line 213, in __init__\n    self.process()\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/seidel.py\", line 230, in process\n    traps = self.query_graph.follow_edge(edge)\n  File \"/home/christian/Documents/dev/udrop/microdrop/microdrop/svg_model/seidel.py\", line 508, in follow_edge\n    while(edge.q.x > trapezoids[-1].right_point.x):\n}}}", 
    "reporter": "cfobel", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1397763541728826", 
    "component": "microdrop", 
    "summary": "Device fails to import", 
    "priority": "major", 
    "keywords": "", 
    "version": "0.1", 
    "time": "2012-05-30T02:30:57", 
    "milestone": "Microdrop 1.0", 
    "owner": "cfobel", 
    "type": "defect"
}
```

**Comment by cfobel** [comment]
Trac update at 20120530T02:31:37: 
* cfobel changed attachment from "" to "KihwanLLLE (copy).svg"
* cfobel commented:

> Device SVG causing exception

**Comment by cfobel** [comment]
Trac update at 20120530T02:35:02: cfobel commented:

> By removing path4698 from the file (and replacing it with a clone of an equivalent path), the device may be imported.

**Comment by cfobel** [comment]
Trac update at 20120801T15:46:26: 
* ryan changed resolution from "" to "fixed"
* ryan changed status from "new" to "closed"

**Comment by cfobel** [comment]
Trac update at 20130410T14:01:31: 
* ryan commented:

> There's still a bug in the svg import code. The following test-case demonstrates the issue. Because the seidel.py module randomly shuffles the edges each time, we don't see the error consistently; however, it appears to occur 50% of the time.
> 
> ```
> from seidel import Triangulator
> 
> verts = [(26.090910000000008, 20.543109999999956),
>          (33.597770000000025, 20.543109999999956),
>          (33.597770000000025, 12.413719999999955),
>          (25.46839, 12.413719999999955),
>          (25.46839, 20.451549999999997),
>          (26.090910000000008, 20.451549999999997),
>          (26.090910000000008, 20.54314999999997)]
> 
> fail = 0
> for i in range(0, 1000):
>     try:
>         triangulator = Triangulator(verts)
>     except:
>         fail+=1
> 
> print fail
> ```


* ryan changed resolution from "fixed" to ""
* ryan changed status from "closed" to "reopened"

**Comment by cfobel** [comment]
Trac update at 20130410T16:14:45: ryan commented:

> Filed issue with the seidel.py author:
> 
> http://code.google.com/p/poly2tri/issues/detail?id=72

**Comment by cfobel** [comment]
Trac update at 20130411T16:53:31: 
* ryan@fobel.net commented:

> commit 6c57341b9f65502c2d248a11ab9b5c28a97ecd29
> Author: Ryan Fobel <ryan@fobel.net>
> Date:   Thu Apr 11 12:34:12 2013 -0400
> 
>     Fixes #110


* ryan@fobel.net changed resolution from "" to "fixed"
* ryan@fobel.net changed status from "reopened" to "closed"

**Comment by cfobel** [comment]
Trac update at 20140417T19:39:01: ryan changed component from "" to "microdrop"
